### PR TITLE
fixes #gh-385

### LINKED ISSUE #385: Config server yml or properties endpoints can produce unexpected results when merging collections and maps
Please review the attached YAML file for sample properties. Note the default profile has an array of PromoCode objects defined. When a SpringBoot application is configured to talk to the Config Server with the appropriate profile, you get the correct response. However, when you request the properties or yaml view, you get a mixed array of the 2 profiles.

    http://localhost:8888/promoCodes/default - produces correct json
    http://localhost:8888/promoCodes/dev - produces correct json

Alternate views produce correct results for default profile

    http://localhost:8888/promoCodes-default.properties

    promos.promoCards[0].client: Google Nexus 6P     <----
    promos.promoCards[0].description: Free stuff for Google users     <----
    promos.promoCards[0].endDate: 2016-04-26T20:01:30.781Z     <----
    promos.promoCards[0].image: http://google.com/images/coolpic.jpg     <----
    promos.promoCards[0].name: P1     <----
    promos.promoCards[0].platform: Android     <----
    promos.promoCards[0].startDate: 2016-04-26T20:01:30.781Z     <----
    promos.promoCards[0].thumbnail: http://google.com/images/coolpic.jpg     <----
    promos.promoCards[1].client: Google Nexus 5X
    promos.promoCards[1].description: Free stuff for Google users
    promos.promoCards[1].endDate: 2016-04-26T20:01:30.781Z
    promos.promoCards[1].image: http://google.com/images/coolpic.jpg
    promos.promoCards[1].name: P2
    promos.promoCards[1].platform: Android
    promos.promoCards[1].startDate: 2016-04-26T20:01:30.781Z
    promos.promoCards[1].thumbnail: http://google.com/images/coolpic.jpg
    promos.promoCards[2].client: iPhone 6S
    promos.promoCards[2].description: Expensive stuff from Steve
    promos.promoCards[2].endDate: 2016-04-26T20:01:30.781Z
    promos.promoCards[2].image: http://apple.com/images/coolpic.jpg
    promos.promoCards[2].name: P3
    promos.promoCards[2].platform: iOS 9
    promos.promoCards[2].startDate: 2016-04-26T20:01:30.781Z
    promos.promoCards[2].thumbnail: http://apple.com/images/coolpic.jpg
    promos.promoCards[3].client: Nokia Lumia 8610
    promos.promoCards[3].description: Bad stuff for windoze users
    promos.promoCards[3].endDate: 2016-04-26T20:01:30.781Z
    promos.promoCards[3].image: http://microsoft.com.com/images/coolpic.jpg
    promos.promoCards[3].name: P4
    promos.promoCards[3].platform: Windoze
    promos.promoCards[3].startDate: 2016-04-26T20:01:30.781Z
    promos.promoCards[3].thumbnail: http://microsoft.com.com/images/coolpic.jpg

Alternate views produce incorrect results for dev profile

    http://localhost:8888/promoCodes-dev.properties

    promos.promoCards[0].client: Nokia Lumia 8888     <----
    promos.promoCards[0].description: Bad stuff for windoze users     <----
    promos.promoCards[0].endDate: 2016-04-26T20:01:30.781Z     <----
    promos.promoCards[0].image: http://microsoft.com.com/images/coolpic.jpg     <----
    promos.promoCards[0].name: P4     <----
    promos.promoCards[0].platform: Windoze     <----
    promos.promoCards[0].startDate: 2016-04-26T20:01:30.781Z     <----
    promos.promoCards[0].thumbnail: http://microsoft.com.com/images/coolpic.jpg     <----
    promos.promoCards[1].client: Google Nexus 5X
    promos.promoCards[1].description: Free stuff for Google users
    promos.promoCards[1].endDate: 2016-04-26T20:01:30.781Z
    promos.promoCards[1].image: http://google.com/images/coolpic.jpg
    promos.promoCards[1].name: P2
    promos.promoCards[1].platform: Android
    promos.promoCards[1].startDate: 2016-04-26T20:01:30.781Z
    promos.promoCards[1].thumbnail: http://google.com/images/coolpic.jpg
    promos.promoCards[2].client: iPhone 6S
    promos.promoCards[2].description: Expensive stuff from Steve
    promos.promoCards[2].endDate: 2016-04-26T20:01:30.781Z
    promos.promoCards[2].image: http://apple.com/images/coolpic.jpg
    promos.promoCards[2].name: P3
    promos.promoCards[2].platform: iOS 9
    promos.promoCards[2].startDate: 2016-04-26T20:01:30.781Z
    promos.promoCards[2].thumbnail: http://apple.com/images/coolpic.jpg
    promos.promoCards[3].client: Nokia Lumia 8610
    promos.promoCards[3].description: Bad stuff for windoze users
    promos.promoCards[3].endDate: 2016-04-26T20:01:30.781Z
    promos.promoCards[3].image: http://microsoft.com.com/images/coolpic.jpg
    promos.promoCards[3].name: P4
    promos.promoCards[3].platform: Windoze
    promos.promoCards[3].startDate: 2016-04-26T20:01:30.781Z
    promos.promoCards[3].thumbnail: http://microsoft.com.com/images/coolpic.jpg

 Note the difference in promos.promoCards[0] between the 2 results indicated by <----.

The desired output should be for the dev profile:

    http://localhost:8888/promoCodes-dev.properties

    promos.promoCards[0].client: Nokia Lumia 8888
    promos.promoCards[0].description: Bad stuff for windoze users
    promos.promoCards[0].endDate: 2016-04-26T20:01:30.781Z
    promos.promoCards[0].image: http://microsoft.com.com/images/coolpic.jpg
    promos.promoCards[0].name: P4
    promos.promoCards[0].platform: Windoze
    promos.promoCards[0].startDate: 2016-04-26T20:01:30.781Z
    promos.promoCards[0].thumbnail: http://microsoft.com.com/images/coolpic.jpg

The issue occurs in org.springframework.cloud.config.server.environment.EnvironmentController, specifically convertToProperties(Environment environment):

	private Map<String, Object> convertToProperties(Environment profiles) {
		Map<String, Object> map = new TreeMap<>();
		List<PropertySource> sources = new ArrayList<>(
				profiles.getPropertySources());
		Collections.reverse(sources);
		for (PropertySource source : sources) {
			@SuppressWarnings("unchecked")
			Map<String, String> value = (Map<String, String>) source.getSource();
			map.putAll(value);                               <-- (No checking for existing array from earlier profile with same name)
		}
		postProcessProperties(map);
		return map;
	}

The issue is the for loop over the PropertySources. The Map<String, String> value has a key that contains all the individual promos.promoCards[X] entries individually and puts them all on each iteration into a final map, replacing any duplicates.  However, in the case of a YAML list, it should also be checking if any other 'promos.promoCards[' exists as keys and remove them before adding all of the current iteration to the resulting map.

[promoCodes.yml.txt](https://github.com/spring-cloud/spring-cloud-config/files/241206/promoCodes.yml.txt)

### PR DESCRIPTION
**Pull request by dmfrey**
Added correct handling of arrays/maps in properties and yaml config server endpoints

### CODE REVIEW

File: spring-cloud-config-server/src/test/java/org/springframework/cloud/config/server/environment/EnvironmentControllerTests.java [hunk 1]
```diff
@@ -98,6 +99,28 @@ public void arrayInYaml() throws Exception {
...
+		oneMap.put("a.b[0]", "c");
+		oneMap.put("a.b[1]", "d");
+		oneMap.put("a.b[2]", "z");
+		this.environment.add(new PropertySource("one", oneMap));
+
+		Mockito.when(this.repository.findOne("foo", "bar", null)).thenReturn(this.environment);
+		String yaml = this.controller.yaml("foo", "bar", false).getBody();
+
+		// Result will not contain original, extra values from oneMap
+		assertEquals("a:\n  b:\n  - f\n  - h\n", yaml);
```
Comment (dsyer) [inline spring-cloud-config-server/src/test/java/org/springframework/cloud/config/server/environment/EnvironmentControllerTests.java]: I feel like we need an assertion about the content of the `Environment` as well to make sure it is consistent with the YAML.

Comment (dmfrey) [inline spring-cloud-config-server/src/test/java/org/springframework/cloud/config/server/environment/EnvironmentControllerTests.java]: @dsyer Should I add a test similar to others in there like the following:

	@Test
	public void mappingForYaml() throws Exception {
		Mockito.when(this.repository.findOne("foo", "bar", null)).thenReturn(this.environment);
		MockMvc mvc = MockMvcBuilders.standaloneSetup(this.controller).build();
		mvc.perform(MockMvcRequestBuilders.get("/foo-bar.yml"))
		.andExpect(
				MockMvcResultMatchers.content().contentType(MediaType.TEXT_PLAIN))
				.andExpect(MockMvcResultMatchers.content().string("{}\n"));
	}

Is that what you are referring to? If so, I can add another test in there for that.

Comment (dsyer) [inline spring-cloud-config-server/src/test/java/org/springframework/cloud/config/server/environment/EnvironmentControllerTests.java]: No, actually: I meant that the values you get from the Spring `Environment` in a config client should be consistent with the values you get from the controller's YAML and properties endpoints. You don't have to be a config client to test that, just build an `Environment` from the output of the `EnvironmentController.labelled()` and assert the contents.

Comment (dmfrey) [inline spring-cloud-config-server/src/test/java/org/springframework/cloud/config/server/environment/EnvironmentControllerTests.java]: Ah, I see. I misunderstood. I will add that in there and issue an update to the PR.

Comment (dmfrey) [inline spring-cloud-config-server/src/test/java/org/springframework/cloud/config/server/environment/EnvironmentControllerTests.java]: @dsyer Are you referring to a Spring Cloud Environment or Spring Core Environment? The Spring Cloud Environment contains the property sources, not the resolved properties per the profile selected. I can inspect the JSON returned and see the property sources ordered correctly. Is there some other test you are looking for?

Comment (spencergibb) [inline spring-cloud-config-server/src/test/java/org/springframework/cloud/config/server/environment/EnvironmentControllerTests.java]: He means the `Environment` returned from `EnvironmentController.labelled()`. No need for mock mvc and json.

Comment (dsyer) [inline spring-cloud-config-server/src/test/java/org/springframework/cloud/config/server/environment/EnvironmentControllerTests.java]: Actually I meant both - to assert compatibility I think you need to create a Spring Core `Environment` and assert it has the same properties.

Comment (dmfrey) [inline spring-cloud-config-server/src/test/java/org/springframework/cloud/config/server/environment/EnvironmentControllerTests.java]: I have been looking through the test code for the server as well as the client and the doesn't appear to be any tests that convert the Spring Cloud Config Environment to a Spring Core Environment. Do you have a standard you would like to follow?

Comment (dsyer) [inline spring-cloud-config-server/src/test/java/org/springframework/cloud/config/server/environment/EnvironmentControllerTests.java]: The config client does it. That code is in the classpath of the server, but I guess it doesn't have to be shared, if you feel like writing or copying a few lines of it.

Comment (dmfrey) [inline spring-cloud-config-server/src/test/java/org/springframework/cloud/config/server/environment/EnvironmentControllerTests.java]: @dsyer Can you point me to an existing test case to look at? I have been going through the spring-cloud-config-client test cases. I am not seeing any that would go through the code that I modified. Some mock the cloud environment and some mock the core environment. Thanks.

BTW, I will be joining Pivotal on June 6th.

Comment (dsyer) [inline spring-cloud-config-server/src/test/java/org/springframework/cloud/config/server/environment/EnvironmentControllerTests.java]: I don't know of an existing test that does this, sorry. Welcome aboard (nearly).

**Comment by dsyer** [comment]
Closed in 23238

**Pull request closed without merging**